src/markdown/tutorial/part-2/11-ember-data.md

```diff
@@ -251,60 +253,95 @@ Darn, there were a couple of failing tests! At the same time, it's great that we
+Let's start customizing the things that didn't work for us by default. Specifically:
 
-The idea is that, provided that your backend exposes a *consistent* protocol and interchange format to access its data, we can write a single adapter-serializer pair to handle all data fetches for the entire application.
+- Our resource URLs have an extra `/api` *namespace* prefix.
+- Our resource URLs have a `.json` extension at the end.
 
-As it turns out, JSON:API just happens to be EmberData's default data protocol and interchange format. Out of the box, EmberData provides a default JSON:API adapter and serializer. This is great news for us, since that is also what our server has implemented. What a wonderful coincidence!
+The first thing we want to do is have our builder respect a configurable default host and/or namespace. Adding a namespace prefix happens to be pretty common across Ember apps, so EmberData provides a global config mechanism for host and namespace. Typically you will want to do this either in your store file or app file.
 
-However, as mentioned above, there are some minor differences between how our server works and EmberData's default assumptions. We can customize the default behavior by defining our own adapter and serializer:
+```run:file:patch lang=js cwd=super-rentals filename=app/app.js
+@@ -2,6 +2,11 @@
+ import Application from '@ember/application';
+ import Resolver from 'ember-resolver';
+ import loadInitializers from 'ember-load-initializers';
+ import config from 'super-rentals/config/environment';
++import { setBuildURLConfig } from '@ember-data/request-utils';
++
++setBuildURLConfig({
++  namespace: 'api',
++});
 
-```run:file:create lang=js cwd=super-rentals filename=app/adapters/application.js
-import JSONAPIAdapter from '@ember-data/adapter/json-api';
+ export default class App extends Application {
+   modulePrefix = config.modulePrefix;
+```
 
-export default class ApplicationAdapter extends JSONAPIAdapter {
-  namespace = 'api';
+Adding the `.json` extension is a bit less common, and doesn't have a declarative configuration API of its own. We could just modify request options directly in place of use, but that would be a bit messy. Instead, let's create a handler to do this for us.
 
-  buildURL(...args) {
-    return `${super.buildURL(...args)}.json`;
-  }
-}
-```
+```run:file:create cwd=super-rentals filename=app/utils/handlers.js
+const JsonSuffixHandler = {
+  request(context, next) {
+    const { request } = context;
+    request.url += '.json';
```

Comment:
should request be immutable? (or it's response only?)

Reply:
request is indeed immutable, this needs to be:

```ts
const updatedRequest = Object.assign({}, request, { url: request.url + '.json' });
```

Reply:
I'll change, but this worked fine. Should I add an assert of freeze `url` somehow on request object?

Reply:
it is frozen in prod build